### ISSUE TITLE
Loss computation for self-supervised datasets with no projection in the model

### ISSUE THREAD

**Issue by Ahmed-Khaled-Saleh** [rbalestr-lab/lejepa#32]
Thanks for the interesting work. @RandallBalestriero 

I am trying to replicate lejepa in a self-supervised setting with no labels but I am a little confused about passing `proj` or `embed` to segref loss and inv_loss. In the [minimal](https://github.com/rbalestr-lab/lejepa/blob/main/MINIMAL.md) example, you assume you get a projection:

```emb, proj = net(vs)
inv_loss = (proj.mean(0) - proj).square().mean()
sigreg_loss = sigreg(proj)
lejepa_loss = sigreg_loss * cfg.lamb + inv_loss * (1 - cfg.lamb)
```

then you pass emb to the linear probing layer: `probe(emb.detach())` not `probe(proj.detach())`.

In my case, I only have an embedding getting out of the model. My model architecture (LeNet-style) is set up to output a vector representation of size, say 512 (embedding). Per the paper's algorithm 2, I can do the following:

```
g_embed = forward(views)
a_embed = g_embed.detach().clone() # cnn-like should have the g_embed = a_embed as per the paper. 

a_embed = a_embed.view(-1, bs, K)
centers = g_embed.view(-1, bs, K).mean(0)
sim = (centers - a_embed ).square().mean()
total_loss= (1-lambda) * sim + lambda * mean(SigReg(emb, global_step) for embed in a_embed))

```

So, my question is, why in the minimal example, the projection output is the one used instead of the embedding as opposed to algorithm 2 of the paper? And if we consider the projector output as the embedding, why it's not used for the probing?

Also, do you think that lejepa can work for small images (say $40 \times 40$ or even $32 \times 32$) with LeNet-style architectures?

**Comment by RandallBalestriero** [comment]
Hello @Ahmed-Khaled-Saleh !

Great question, the use of a projector is a pretty established technique in SSL (see https://arxiv.org/abs/2206.13378 for example). It allows to gain a few percentage points by reducing the need to thoroughly cross-validate data-augmentations and other hyper-parameters. The idea is to plug the output of the projector into your SSL loss, but the output of your encoder to the linear probe. Without a projector, performances will likely be a bit lower (we have some ablations in the paper), but I would recommend just sticking a 2~3 layer MLP as a projector and reuse our minimal setup (no matter which encoder you are using).

Regarding small images, yes I know a few researchers who have made it work with resnet18 on cifar100, so it definitely will work! Closing this but feel free to ping back if this didn't answer your question.
